include/openmc/random_ray/source_region.h

```diff
@@ -81,11 +81,211 @@ struct TallyTask {
   };
 };
 
+// The SourceRegionKey combines a base source region (i.e., a material
+// filled cell instance) with a mesh bin. This key is used as a handle
+// for dynamically discovered source regions when subdividing source
+// regions with meshes.
+class SourceRegionKey {
+public:
+  int64_t base_source_region_id;
+  int64_t mesh_bin;
+  SourceRegionKey() = default;
+  SourceRegionKey(int64_t source_region, int64_t bin)
+    : base_source_region_id(source_region), mesh_bin(bin)
+  {}
+
+  // Equality operator required by the unordered_map
+  bool operator==(const SourceRegionKey& other) const
+  {
+    return base_source_region_id == other.base_source_region_id &&
+           mesh_bin == other.mesh_bin;
+  }
+
+  // Hashing functor required by the unordered_map
+  struct HashFunctor {
+    size_t operator()(const SourceRegionKey& key) const
+    {
+      size_t seed = 0;
+      hash_combine(seed, key.base_source_region_id);
+      hash_combine(seed, key.mesh_bin);
+      return seed;
+    }
+  };
+};
+
+class SourceRegionHandle {
+public:
+  //----------------------------------------------------------------------------
+  // Public Data members
+  int negroups_;
+  bool is_numerical_fp_artifact_ {false};
+  bool is_linear_ {false};
+
+  // Scalar fields
+  int* material_;
+  int* is_small_;
+  int* n_hits_;
+  int* birthday_;
+  OpenMPMutex* lock_;
+  double* volume_;
+  double* volume_t_;
+  double* volume_sq_;
+  double* volume_sq_t_;
+  double* volume_naive_;
+  int* position_recorded_;
+  int* external_source_present_;
+  Position* position_;
+  Position* centroid_;
+  Position* centroid_iteration_;
+  Position* centroid_t_;
+  MomentMatrix* mom_matrix_;
+  MomentMatrix* mom_matrix_t_;
+  // A set of volume tally tasks. This more complicated data structure is
+  // convenient for ensuring that volumes are only tallied once per source
+  // region, regardless of how many energy groups are used for tallying.
+  std::unordered_set<TallyTask, TallyTask::HashFunctor>* volume_task_;
+
+  // Mesh that subdivides this source region
+  int* mesh_;
+  int64_t* parent_sr_;
+
+  // Energy group-wise 1D arrays
+  double* scalar_flux_old_;
+  double* scalar_flux_new_;
+  double* source_;
+  double* external_source_;
+  double* scalar_flux_final_;
+
+  MomentArray* source_gradients_;
+  MomentArray* flux_moments_old_;
+  MomentArray* flux_moments_new_;
+  MomentArray* flux_moments_t_;
+
+  // 2D array representing values for all energy groups x tally
+  // tasks. Each group may have a different number of tally tasks
+  // associated with it, necessitating the use of a jagged array.
+  vector<TallyTask>* tally_task_;
+
+  //----------------------------------------------------------------------------
+  // Public Accessors
+
+  int& material() { return *material_; }
+  const int& material() const { return *material_; }
```

Comment:
For the `const` accessors, I think it's preferable to just return by value rather than a const reference.

Reply:
Will do! For the fields that involve dynamic memory allocation, I am planning to leave them as references.